### DISCUSSION TITLE
Stuck on chat privacy (realtime with private user ids)

### DISCUSSION THREAD

**Discussion by EskelCz** [supabase/supabase#15332] [Questions] [answered]
I need some help figuring this out, I feel like I ran out of options.

I'm building a chat feature in my app, which should allow each user to have multiple profiles and protect their actual user id.

The messages have basically: `text:varchar, profileId:uuid, conversation_id:uuid, created:timestampz`

I was hoping to keep the conversation_id secret just between the chat participants. To restrict `SELECT *` and only allow `WHERE conversation_id`. Therefore I wouldn't have to have a complicated and inefficient RLS policy with joins, to try to figure out which users can read which messages. But I can't figure out how to actually do that, with realtime. :(

I tried:
1. Using an RPC `get_messages` with `conversation_id` param. That's great for loading of previous messages, but not realtime (new messages).
2. `(POSITION(('WHERE'::text) IN (current_query())) > 0)` to check for WHERE clause in the query. Again, doesn't work with realtime.
3. Denormalize the table with a trigger on adding a new message (save participant ids) and use a view to return messages without the id columns. Then I found views don't work with realtime either.

I'm expecting a large number of messages being stored in the table.

Can this be done?

**Comment by EskelCz** [comment, marked as answer]
Finally found a way. It's possible to revoke column privileges which can make certain columns private. It's a postgres feature and works with realtime just fine, the columns are simply omitted. Just what I needed.

**Reply by NotHolst** [reply]
Could you possibly post an example on how you did this? I've tried
```sql
REVOKE SELECT (name) ON "public"."projects" from public;
```
But it sadly didn't work

**Reply by EskelCz** [reply]
@NotHolst  Try this
```
REVOKE ALL PRIVILEGES ON projects FROM public;
GRANT SELECT (name) ON projects TO public;
```
